File: 0322-coin-change/solution.py
class Solution(object):
    
    def minCoins(self, coins, amount, dp):
        if amount in dp:
            return dp[amount]
        
        if amount < 0:
            dp[amount] = -1
            return -1
        
        if amount == 0:
            dp[amount] = 0
            return 0

        minimumCoins = float('inf')
        for coin in coins:
            res = self.minCoins(coins, amount-coin, dp) + 1
            if res > 0 and res < minimumCoins:
                minimumCoins = res
        
        if minimumCoins == float('inf'):
            minimumCoins = -1
        dp[amount] = minimumCoins
        return minimumCoins
                
        
        

    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = dict()
        return self.minCoins(coins, amount, dp)

File: 0322-coin-change/solution.cpp
class Solution {
public:
    int helper(vector<int>& coins, int amount, unordered_map<int, int> & memo) {
        if (amount == 0) return 0;
        else if (amount < 0) return -1;
        
        if (memo.find(amount) != memo.end()) return memo[amount];
        
        int ans = INT_MAX;
        for (auto coin : coins) {
            int numCoinsRequired = helper(coins, amount - coin, memo);
            if (numCoinsRequired == -1) {
                continue;
            }
            ans = min(ans,  numCoinsRequired + 1);
        }        

        if (ans == INT_MAX) ans = -1;
        memo[amount] = ans;
        return ans;

    }
    int coinChange(vector<int>& coins, int amount) {
        unordered_map<int, int> memo;
        return helper(coins, amount, memo);
    }
};
